frame/contracts/src/lib.rs

```diff
@@ -967,6 +989,156 @@ struct InternalOutput<T: Config, O> {
 	result: Result<O, ExecError>,
 }
 
+/// Helper trait to wrap contract execution entry points into a signle function
+/// [`Invokable::run_guarded`].
+trait Invokable<T: Config, O> {
+	/// Single entry point to contract execution.
+	/// Downstream execution flow is branched by implementations of [`Invokable`] trait:
+	///
+	/// - [`InstantiateInput::run`] runs contract instantiation,
+	/// - [`CallInput::run`] runs contract call.
+	///
+	/// We enforce a re-entrancy guard here by initializing and checking a boolean flag through a
+	/// global reference.
+	fn run_guarded(&self, context: ExecContext<T>) -> InternalOutput<T, O> {
+		let gas_limit = context.gas_limit;
+		executing_contract::using_once(&mut false, || {
+			executing_contract::with(|f| {
+				// Fail if already entered contract execution
+				if *f {
+					return Err(())
+				}
+				// We are entering contract execution
+				*f = true;
+				Ok(())
+			})
+			.unwrap_or(Ok(())) // Should always be `Some`, still we make it safe.
+			.map_or_else(
+				|_| InternalOutput {
+					gas_meter: GasMeter::new(gas_limit),
+					storage_deposit: Default::default(),
+					result: Err(ExecError {
+						error: <Error<T>>::ReentranceDenied.into(),
+						origin: ErrorOrigin::Caller,
+					}),
+				},
+				// Enter contract call.
+				|_| self.run(context, GasMeter::new(gas_limit)),
+			)
+		})
```

Comment:
I would prefer two calls to `map` and `map_err`. It might be a preference but I think it makes it more clear which of those is called on error and which is called on ok.

Reply:
This would require unwrapping the Result, which is kinda ugly, being unnecessary. I would leave it as is then.

Reply:
Yes you are right.